rs/sns/cli/src/upgrade_sns_controlled_canister.rs

```diff
@@ -0,0 +1,609 @@
+        .await
+        .context("Cannot read canister controllers.")?;
+
+    let cbor: Value = serde_cbor::from_slice(&controllers_blob)
+        .expect("Invalid cbor data for controller controllers.");
+
+    let Value::Array(controllers) = cbor else {
+        bail!("Expected controllers to be an array, but got {cbor:?}");
+    };
+
+    let (controllers, errors): (BTreeSet<_>, Vec<_>) =
+        controllers.into_iter().partition_map(|value| {
+            let Value::Bytes(bytes) = value else {
+                let err =
+                    format!("Expected canister controller to be of type bytes, got {value:?}",);
+                return Either::Right(err);
+            };
+            match Principal::try_from(&bytes) {
+                Err(err) => {
+                    let err = format!("Cannot interpret canister controller principal: {err}");
+                    Either::Right(err)
+                }
+                Ok(principal) => Either::Left(PrincipalId(principal)),
+            }
+        });
+
+    if !errors.is_empty() {
+        let err = format!(
+            "Problems with canister controllers:\n  - {}",
+            errors.join("\n  - ")
+        );
+        bail!(err);
+    }
+
+    Ok((controllers, module_hash))
+}
+
+/// Attempts to create an empty canister on the same subnet as `next_to`.
+///
+/// Returns the ID of the newly created canister in the Ok result.
+pub async fn create_canister_next_to(
+    agent: &Agent,
+    next_to: CanisterId,
+    controllers: Vec<PrincipalId>,
+    cycles_amount: u128,
+    name: &str,
+) -> Result<CanisterId> {
+    // This is expected to be `None` if we're running against a local replica.
+    let subnet_selection = nns::registry::get_subnet_for_canister(agent, next_to)
+        .await
+        .map(|subnet| SubnetSelection::Subnet { subnet })
+        .ok();
+
+    let canister_id = cycles_ledger_create_canister(
+        agent,
+        cycles_amount,
+        subnet_selection,
+        Some(CanisterSettingsArgs {
+            controllers: Some(BoundedVec::new(controllers)),
+            ..Default::default()
+        }),
+    )
+    .await
+    .map_err(|err| {
+        if let CreateCanisterError::InsufficientFunds { balance } = err {
+            let err = format!(
+                "Requested creating the {} canister with {} cycles, but the caller identity has \
+                 only {} cycles on the cycles ledger. Please buy more cycles using \
+                 `dfx cycles convert --amount AMOUNT --network NETWORK` and try again.",
+                name, cycles_amount, balance,
+            );
+            anyhow::anyhow!(err)
+        } else {
+            anyhow::anyhow!(format!("{:?}", err))
+        }
+    })?
+    .canister_id;
+
+    CanisterId::try_from_principal_id(canister_id).map_err(|err| anyhow::anyhow!(err))
+}
+
+pub async fn exec(args: UpgradeSnsControlledCanisterArgs, agent: &Agent) -> Result<()> {
+    // Prepare.
```

Comment:
Did you mean to leave this comment here? I don't see any others in this function that seem similar

Reply:
> Did you mean to leave this comment here?

Yes

> I don't see any others in this function

Other comments were turned into prints.